sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -458,7 +458,16 @@ public BundleFinalizer getBundleFinalizer() {
             startFunction.run();
           }
 
-          if (!bundleProcessor.getInboundEndpointApiServiceDescriptors().isEmpty()) {
+          if (request.getProcessBundle().hasElements()) {
+            boolean inputFinished =
+                bundleProcessor
+                    .getInboundObserver()
+                    .multiplexElements(request.getProcessBundle().getElements());
+            if (!inputFinished) {
+              throw new RuntimeException(
+                  "Elements embedded in ProcessBundleRequest are incomplete.");
```

Comment:
nit: It would be nice to enumerate the missing ones in addition to the messaging improvement.

```suggestion
                  "Elements embedded in ProcessBundleRequest do not contain stream terminators for all data and timer inputs.");
```